### PR TITLE
Refactor enum_literal compls

### PR DESCRIPTION
**Pull request by nullptrdevs**
Not perfect, but functional .

### CODE REVIEW

File: src/features/completions.zig [hunk 1]
```diff
@@ -1495,11 +1007,654 @@ pub fn completionAtIndex(server: *Server, analyser: *Analyser, arena: std.mem.Al
...
+                arena,
+                Analyser.TypeWithHandle{
+                    .handle = either.handle,
+                    .type = .{
+                        .data = .{ .other = enode },
+                        .is_type_val = true,
+                    },
+                },
+            );
+        }
```
Comment (Techatrix) [inline src/features/completions.zig:1601]:
```suggestion
        try node_type.getAllTypesWithHandlesArrayList(arena, types_with_handles);
```

File: src/features/completions.zig [hunk 2]
```diff
@@ -1495,11 +1007,654 @@ pub fn completionAtIndex(server: *Server, analyser: *Analyser, arena: std.mem.Al
...
+                        Analyser.TypeWithHandle{
+                            .handle = either.handle,
+                            .type = .{
+                                .data = .{ .other = node },
+                                .is_type_val = true,
+                            },
+                        },
+                    );
+                }
+            }
```
Comment (Techatrix) [inline src/features/completions.zig:1456]:
```suggestion
            const node_type = try analyser.resolveTypeOfNode(.{ .node = type_expr_node, .handle = symbol_decl.handle }) orelse return;
            try node_type.getAllTypesWithHandlesArrayList(arena, types_with_handles);
```

File: src/features/completions.zig [hunk 3]
```diff
@@ -1495,11 +1007,654 @@ pub fn completionAtIndex(server: *Server, analyser: *Analyser, arena: std.mem.Al
...
+            symbol_decl.handle,
+            offsets.nodeToLoc(symbol_decl.handle.tree, param.type_expr).end,
+            dot_context,
+        );
+        for (param_rcts) |prct| try types_with_handles.append(arena, prct);
+        return;
+    }
+    switch (nodes_tags[type_expr_node]) {
+        // decl is `const Alias = @import("MyStruct.zig");`
+        .builtin_call_two => {
```
Comment (Techatrix) [inline src/features/completions.zig:1362]: is this special case necessary? shouldn't the else case already handle this?

File: src/features/completions.zig [hunk 4]
```diff
@@ -1495,11 +1007,654 @@ pub fn completionAtIndex(server: *Server, analyser: *Analyser, arena: std.mem.Al
...
+
+        const param_rcts = try collectContainerNodes(
+            document_store,
+            analyser,
+            arena,
+            symbol_decl.handle,
+            offsets.nodeToLoc(symbol_decl.handle.tree, param.type_expr).end,
+            dot_context,
+        );
+        for (param_rcts) |prct| try types_with_handles.append(arena, prct);
```
Comment (Techatrix) [inline src/features/completions.zig:1357]:
```suggestion
        var fn_param_decl = Analyser.Declaration{ .param_payload = .{
            .func = type_expr_node,
            .param_index = @intCast(dot_context.fn_arg_index),
        } };
        const fn_param_decl_with_handle = Analyser.DeclWithHandle{ .decl = &fn_param_decl, .handle = symbol_decl.handle };
        const param_type = try fn_param_decl_with_handle.resolveType(analyser) orelse return;
        try types_with_handles.append(arena, param_type);
```

File: src/features/completions.zig [hunk 5]
```diff
@@ -1495,11 +1007,654 @@ pub fn completionAtIndex(server: *Server, analyser: *Analyser, arena: std.mem.Al
...
+                    arena,
+                    Analyser.TypeWithHandle{
+                        .handle = either.handle,
+                        .type = .{
+                            .data = .{ .other = node },
+                            .is_type_val = true,
+                        },
+                    },
+                );
+            }
```
Comment (Techatrix) [inline src/features/completions.zig:1340]:
```suggestion
            try node_type.getAllTypesWithHandlesArrayList(arena, types_with_handles);
```

File: src/features/completions.zig [hunk 6]
```diff
@@ -1495,11 +1007,654 @@ pub fn completionAtIndex(server: *Server, analyser: *Analyser, arena: std.mem.Al
...
+/// Resolves `identifier`/`path.to.identifier` at `text_index`
+/// If the `identifier` is a container `fn_arg_index` is unused
+/// If the `identifier` is a `fn_name`/`identifier.fn_name`, tries to resolve
+///         `fn_name`'s `fn_arg_index`'s param type
+fn collectContainerNodes(
+    document_store: *DocumentStore,
+    analyser: *Analyser,
+    arena: std.mem.Allocator,
+    handle: *const DocumentStore.Handle,
+    text_index: usize,
```
Comment (Techatrix) [inline src/features/completions.zig:1266]: `source index` is the term used to refer to byte indices into document source code throughout the codebase.
```suggestion
    source_index: usize,
```

File: src/features/completions.zig [hunk 7]
```diff
@@ -1495,11 +1007,654 @@ pub fn completionAtIndex(server: *Server, analyser: *Analyser, arena: std.mem.Al
...
+    for (containers) |container| {
+        const node = switch (container.type.data) {
+            .other => |n| n,
+            else => continue,
+        };
+        var buffer: [2]Ast.Node.Index = undefined;
+        const container_decl = Ast.fullContainerDecl(container.handle.tree, &buffer, node) orelse continue;
+        for (container_decl.ast.members) |member| {
+            const field = container.handle.tree.fullContainerField(member) orelse continue;
+            if (std.mem.eql(u8, container.handle.tree.tokenSlice(field.ast.main_token), alleged_field_name)) {
```
Comment (Techatrix) [inline src/features/completions.zig:1633]: this can be `if (!std.mem.eql(...)) continue` to avoid extra indentation.

**Review by Techatrix — CHANGES_REQUESTED** [review]
I did not perform any more checks on my suggestions than running tests so they may break something.
The only exception is the suggestion about the `@import` builtin resolution since the current testing infrastructure doesn't support multi file completion tests.

Could you also please remove your debugging related print statements including those that you have comment out.

**Comment by nullptrdevs** [comment]
Applied all of the above, multi file completions seem to work.

File: src/features/completions.zig [hunk 8]
```diff
@@ -1495,11 +1006,499 @@ pub fn completionAtIndex(server: *Server, analyser: *Analyser, arena: std.mem.Al
...
+        return;
+    };
+    const name = offsets.locToSlice(handle.tree.source, name_loc);
+    const decls = try analyser.getSymbolFieldAccesses(arena, handle, loc.end, loc, name) orelse return;
+    for (decls) |decl| {
+        const decl_node = switch (decl.decl.*) {
+            .ast_node => |ast_node| ast_node,
+            else => continue,
+        };
+        var node_type = try analyser.resolveTypeOfNode(.{ .node = decl_node, .handle = decl.handle }) orelse continue;
```
Comment (Techatrix) [inline src/features/completions.zig:1409]:
```suggestion
        var node_type = try decl.resolveType(analyser) orelse continue;
```

File: src/features/completions.zig [hunk 9]
```diff
@@ -1495,11 +1006,499 @@ pub fn completionAtIndex(server: *Server, analyser: *Analyser, arena: std.mem.Al
...
+            const param_rcts = try collectContainerNodes(
+                analyser,
+                arena,
+                container.handle,
+                end,
+                &el_dot_context,
+            );
+            for (param_rcts) |prct| try types_with_handles.append(arena, prct);
+            break;
+        }
```
Comment (Techatrix) [inline src/features/completions.zig:1502]:
```suggestion
        const member_decl = try analyser.lookupSymbolContainer(.{ .node = node, .handle = container.handle }, alleged_field_name, .field) orelse continue;
        const member_type = try member_decl.resolveType(analyser) orelse continue;
        try types_with_handles.append(arena, member_type);
```

File: src/features/completions.zig [hunk 10]
```diff
@@ -1495,11 +1006,499 @@ pub fn completionAtIndex(server: *Server, analyser: *Analyser, arena: std.mem.Al
...
+        // XXX use-case: resolves `if (symbol_decl.decl.* != .`
+        if (container.type.data == .other) {
+            const node_type = try analyser.resolveTypeOfNode(.{ .node = container.type.data.other, .handle = container.handle }) orelse return;
+            if (node_type.isEnumType() or node_type.isUnionType()) {
+                try types_with_handles.append(arena, node_type);
+                return;
+            }
+        }
+    }
+    return;
```
Comment (Techatrix) [inline src/features/completions.zig:1382]:
```suggestion
```

File: src/features/completions.zig [hunk 11]
```diff
@@ -1495,11 +1006,499 @@ pub fn completionAtIndex(server: *Server, analyser: *Analyser, arena: std.mem.Al
...
+                node_loc.end,
+                dot_context,
+            );
+            for (rcts) |rct| try types_with_handles.append(arena, rct);
+        },
+        else => {
+            const node_type = try analyser.resolveTypeOfNode(.{ .node = type_expr_node, .handle = symbol_decl.handle }) orelse return;
+            try node_type.getAllTypesWithHandlesArrayList(arena, types_with_handles);
+        },
+    }
```
Comment (Techatrix) [inline src/features/completions.zig:1344]:
```suggestion
    const type_expr = try symbol_decl.resolveType(analyser) orelse return;
    if (type_expr.isFunc()) {
        const fn_proto_node = type_expr.type.data.other;
        var buf: [1]Ast.Node.Index = undefined;
        const full_fn_proto = symbol_decl.handle.tree.fullFnProto(&buf, fn_proto_node) orelse return;
        if (dot_context.likely == .enum_literal) { // => we need f()'s return type
            if (full_fn_proto.ast.return_type == 0) return;
            const node_type = try analyser.resolveTypeOfNode(.{ .node = full_fn_proto.ast.return_type, .handle = symbol_decl.handle }) orelse return;
            try node_type.getAllTypesWithHandlesArrayList(arena, types_with_handles);
            return;
        }
        var fn_param_decl = Analyser.Declaration{ .param_payload = .{
            .func = fn_proto_node,
            .param_index = @intCast(dot_context.fn_arg_index),
        } };
        const fn_param_decl_with_handle = Analyser.DeclWithHandle{ .decl = &fn_param_decl, .handle = symbol_decl.handle };
        const param_type = try fn_param_decl_with_handle.resolveType(analyser) orelse return;
        try types_with_handles.append(arena, param_type);
        return;
    }
    try type_expr.getAllTypesWithHandlesArrayList(arena, types_with_handles);
```

File: src/features/completions.zig [hunk 12]
```diff
@@ -1495,11 +1006,499 @@ pub fn completionAtIndex(server: *Server, analyser: *Analyser, arena: std.mem.Al
...
+    analyser: *Analyser,
+    arena: std.mem.Allocator,
+    handle: *const DocumentStore.Handle,
+    loc: offsets.Loc,
+    types_with_handles: *std.ArrayListUnmanaged(Analyser.TypeWithHandle),
+) error{OutOfMemory}!void {
+    const held_range = try arena.dupeZ(u8, offsets.locToSlice(handle.tree.source, loc));
+    var tokenizer = std.zig.Tokenizer.init(held_range);
+
+    if (try analyser.getFieldAccessType(handle, loc.end, &tokenizer)) |result| {
```
Comment (Techatrix) [inline src/features/completions.zig:1359]: reduce indentation:
```suggestion
const result = try analyser.getFieldAccessType(handle, loc.end, &tokenizer) orelse return;
```

**Review by Techatrix — CHANGES_REQUESTED** [review]
second round of refactoring and simplification. tests still pass...

**Comment by nullptrdevs** [comment]
Ditto

**Comment by Techatrix** [comment]
LGTM, hopefully nothing has regressed through these changes.
I will run some fuzzing overnight so that we don't introduce any new crashes.

**Review by Techatrix — APPROVED** [review]
Fuzzing only reported a single crash which was unrelated to this PR. (InternPool related)

**Pull request merged**